blockprotocol/blockprotocol
# Fix problematic typings in `Subgraph`

## 🌟 What is the purpose of this PR?

The `Subgraph` object contains correct information about the keys used as identifiers for `vertices` and `edges`. Namely, that elements are identified by the following pairs 
- `(EntityId, TimestampString)`, 
- `(BaseUri, number)`

In an ideal world, this would encapsulate quite a lot more information about a correctly formed `Subgraph`. It would greatly reduce the number of type guards that would need to be called in TS, for example (`isEntityVertex`), as the code path would know if it's dealing with an ontology element or a knowledge-graph one.

Unfortunately, the `number` type confuses the heck out of TypeScript and it actually can cause code to be completely incorrect at the worst of times.

This is because when the `number` type is used as a key in an object, it is inferred as being equivalent to the string `` `${number}` `` which overlaps the string type used for entity related things. But when seeing a plain string index, TS will assume it's an entity-related thing, rather than correctly inferring it could be either as it _could_ overlap the `number` string.

As part of this PR, I investigated trying to have keys that completely do not overlap (either through additional template strings, or some form of type branding), but eventually needed to opt for removing the helpful type information. 

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203543026679380/1203904064506712/f) _(internal)_

## 🚫 Blocked by

- [x] #931 

## 🔍 What does this change?

As usual, see commits, I will open comments directly on code that needs specific attention.

## 📜 Does this require a change to the docs?

These changes are mostly internal, and as we haven't published the `Subgraph` type yet, this _change_ shouldn't need additional documentation besides the in-code one added.

## ⚠️ Known issues

Will open comments on specific parts of the code which posed known issues, otherwise N/A

## 🐾 Next steps

N/A

## 🛡 What tests cover this?

- `tsc`
- `eslint`
- Manually running MBD

## ❓ How to test this?

1.  See CI
2. Checkout the branch and ensure `yarn` works
3. Try running MBD and ensure it all behaves as expected (noting that a few of the blocks are broken on this branch)

## 📹 Demo

<img width="587" alt="image" src="https://user-images.githubusercontent.com/25749103/216953397-5fc3127c-d744-4f57-8cc0-269339f2eef6.png">
<img width="1187" alt="image" src="https://user-images.githubusercontent.com/25749103/216953615-cd107574-d6c1-42ee-8d82-8cc28bb32386.png">
